pkg/engine/mutate/strategicPreprocessing.go

```diff
@@ -14,539 +30,430 @@ import (
-			}
-		}
-	}
-	// remove the elements with anchors
-	err = removeAnchorElements(pattern)
-	if err != nil {
-		return err
-	}
-	return preProcessArrayPattern(pattern, resource)
-}
 
-func preProcessArrayPattern(pattern, resource *yaml.RNode) error {
-	patternElements, err := pattern.Elements()
-	if err != nil {
-		return err
-	}
 	resourceElements, err := resource.Elements()
 	if err != nil {
 		return err
 	}
+
 	for _, patternElement := range patternElements {
-		patternNameField := patternElement.Field("name")
-		if patternNameField != nil {
-			patternNameValue, err := patternNameField.Value.String()
-			if err != nil {
-				return err
-			}
+		// If pattern has conditions, look for matching elements and process them
+		if hasAnchors(patternElement) {
 			for _, resourceElement := range resourceElements {
-				resourceNameField := resourceElement.Field("name")
-				if resourceNameField != nil {
-					resourceNameValue, err := resourceNameField.Value.String()
+				err := preProcessRecursive(logger, patternElement, resourceElement)
+				if err != nil {
+					if _, ok := err.(ConditionError); ok {
+						// Skip element, if condition has failed
+						continue
+					}
+
+					return err
+				} else {
+					// If condition is satisfied, create new pattern list element based on patternElement
+					// but related with current resource element by name.
+					// Resource element must have name. Without name kustomize won't be able to update this element.
+					// In case if element does not have name, skip it.
+					resourceElementName := resourceElement.Field("name")
+					if resourceElementName.IsNilOrEmpty() {
+						continue
+					}
+
+					newNode := patternElement.Copy()
+					err := deleteConditionsFromNestedMaps(newNode)
+					if err != nil {
+						return err
+					}
+
+					err = newNode.PipeE(yaml.SetField("name", resourceElementName.Value))
 					if err != nil {
 						return err
 					}
-					if patternNameValue == resourceNameValue {
-						err := preProcessPattern(patternElement, resourceElement)
-						if err != nil {
-							return err
-						}
+
+					err = pattern.PipeE(yaml.Append(newNode.YNode()))
+					if err != nil {
+						return err
 					}
 				}
 			}
 		}
 	}
+
 	return nil
 }
 
-/*
-	removeAnchorSequence :- removes element containing conditional anchor
-
-	Pattern:
-		"spec": {
-			"containers": [{
-			"(image)": "*:latest",
-			"imagePullPolicy": "Always"
-		},
-		{
-			"name": "nginx",
-			"imagePullPolicy": "Always"
-		}]}
-	After Removing Conditional Sequence:
-		"spec": {
-			"containers": [{
-			"name": "nginx",
-			"imagePullPolicy": "Always"
-		}]}
-*/
-func removeAnchorElements(pattern *yaml.RNode) error {
-	patternElements, err := pattern.Elements()
+// validateConditions checks all conditions from current map.
+// If at least one condition fails, return error.
+// If caller handles list of maps and gets an error, it must skip element.
+// If caller handles map, it must stop processing and skip entire rule.
```

Comment:
Getting confused about these 2 statements, this function is only invoked by `walkMap` and there it returns on error. Is there any case for "list of maps"?